risc0/zkvm/src/host/ipc/server.rs

```diff
@@ -0,0 +1,280 @@
+
+        let kind = reply
+            .kind
+            .ok_or(IoError::new(IoErrorKind::Other, "Malformed message"))?;
+
+        match kind {
+            ClientIoReplyKind::Ok(_) => Ok(buf.len()),
+            ClientIoReplyKind::Error(err) => Err(IoError::new(IoErrorKind::Other, err)),
+        }
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        Ok(())
+    }
+}
+
+#[derive(Clone)]
+struct SliceIoProxy<'a> {
+    channel: Channel<'a>,
+}
+
+impl<'a> SliceIo for SliceIoProxy<'a> {
+    fn handle_io(&mut self, _syscall: &str, _from_guest: Bytes) -> Result<Bytes> {
+        let request = ClientIoRequest::new();
+        self.channel
+            .send(request)
+            .map_err(|err| IoError::new(IoErrorKind::Other, err.to_string()))?;
+
+        Ok(Bytes::new())
+    }
+}
+
+impl IpcServer {
+    /// TODO
+    pub fn run(&self) -> Result<()> {
+        let mut channel = Channel::new(stdin(), stdout());
+
+        let request: ServerRequest = channel.recv()?;
+        match request.kind.ok_or(malformed_err())? {
+            ServerRequestKind::Hello(_version) => {
+                let mut msg = HelloReply::new();
+                msg.set_version(0);
+                channel.send(msg)?;
+            }
+            ServerRequestKind::Execute(request) => {
+                self.on_execute(channel, request)?;
+            }
+            ServerRequestKind::Prove(request) => {
+                self.on_prove(channel, request)?;
+            }
+        };
+
+        Ok(())
+    }
+
+    fn on_execute(&self, mut channel: Channel, request: ServerExecuteRequest) -> Result<()> {
+        let request = request.execute.as_ref().ok_or(malformed_err())?;
+        let asset = request.binary.asset.as_ref().ok_or(malformed_err())?;
+        let bin_type = request
+            .binary
+            .kind
+            .enum_value()
+            .map_err(|x| anyhow!("Unknown BinaryType: {x}"))?;
+        let image = asset.as_image(bin_type)?;
+
+        let mut env_builder = ExecutorEnv::builder();
+        env_builder.env_vars(request.env_vars.clone());
+        for (fd, _) in request.read_fds.iter() {
+            let proxy = PosixIoProxy {
+                fd: *fd,
+                channel: channel.clone(),
+            };
+            let reader = BufReader::new(proxy);
+            env_builder.read_fd(*fd, reader);
+        }
+        for (fd, _) in request.write_fds.iter() {
+            let proxy = PosixIoProxy {
+                fd: *fd,
+                channel: channel.clone(),
+            };
+            env_builder.write_fd(*fd, proxy);
+        }
+        let proxy = SliceIoProxy {
+            channel: channel.clone(),
+        };
+        for (name, _) in request.slice_ios.iter() {
+            env_builder.slice_io(&name, proxy.clone());
+        }
+        let env = env_builder.build()?;
+
+        let mut exec = Executor::new(env, image)?;
+        let session = exec.run_with_callback(|segment| {
```

Comment:
This is a can-of-worms note so consider it whackadoodle:

Would this be a place that it would make sense to make the IPC server async with something like tokio? In other places I have tried to keep everything in the executor callback inside a tokio Task so the main exec loop could continue unblocked by IO (up to a certain limit).

Maybe IPC IO bandwidth is high enough that this wont be an issue in average cases. I would recommend we do some stress testing with zeth because it's segments are generally the largest I encountered so far.

Reply:
Definitely can-of-worms, but something I've been thinking about too. Want to gather more data, zeth is a good call for that.

Ultimately I think we'll want the server to write to disk as fast as it can in the background with multiple threads (just like Bonsai is doing), and send notifications to clients as things get flushed. My rough impression is that in this case async won't buy us much since I think most of the time we're either going to be computing or waiting for disk I/O and I don't know if switching between threads will be that much overhead (as opposed to something like green threads with a good async runtime).

Reply:
agreed. I think well just have to do some data collection to work out what model gets us the best disk IO saturation.